### ISSUE TITLE
Phase 2: Create GEMMA Hyperparameter Tuning Script (GEMMA Optimizasyon Aracını Geliştir)

### ISSUE THREAD

**Issue by SefaGH** [SefaGH/bearish-alpha-bot#372]
### 🎯 Amaç (Goal)

Bu görevin temel amacı, LSTM modeli için başarıyla çalışan optimizasyon formülünü, GEMMA modeline uyarlamaktır. Bu amaçla, `scripts/tune_regime_models_standalone.py` script'ini bir şablon olarak kullanarak, GEMMA'nın MLP (Feed-forward) mimarisi için hiperparametre optimizasyonu yapacak **`scripts/tune_gemma_model_standalone.py`** adında yeni bir script oluşturulacaktır.

Nihai hedef, GEMMA modelinin potansiyelini ortaya çıkaracak en iyi hiperparametre setini, `Optuna` kullanarak sistematik bir şekilde bulmaktır.

### ⚙️ Referans ve Strateji (Reference & Strategy)

*   **Ana Şablon (Blueprint):** Bu görevdeki tüm mantık, veri işleme adımları ve `Optuna` entegrasyonu için temel referans dosyası **`scripts/tune_regime_models_standalone.py`**'dir. Yeni bir mantık icat etme, bu kanıtlanmış formülü kopyala ve uyarla.
*   **Model Mimarisi Referansı:** Oluşturulacak GEMMA modelinin (MLP) katman yapısı için referans alınacak kod, **`scripts/train_all_models.py`** dosyası içindeki `train_gemma_model` fonksiyonunda bulunan `nn.Sequential` bloğudur.

### ⚠️ Ön Koşullar: Gerekli Artifact Dosyaları

Bu göreve başlamadan önce, FAZ 1'de oluşturulan aşağıdaki "artifact" dosyalarının, belirtilen konumlarda mevcut olduğundan emin olunmalıdır. Bu dosyalar, optimizasyon sürecinin doğru veri üzerinde çalışması için zorunludur.

| Dosya Adı | Gerekli Konum |
| :--- | :--- |
| `BTC-USDT_training_data.npz` | `data/cache/` |
| `feature_selection_mask.npy` | `data/cache/` |

**Not:** Bu dosyalar, `full-lstm-tuning.yml` workflow'unun ilk adımlarında (`prepare_training_data` ve `analyze_features`) üretilir. FAZ 1'de bu süreç doğrulanmıştır.

#### **Adım 1: Yeni Optimizasyon Script Dosyasını Oluşturma**

1.  Mevcut `scripts/tune_regime_models_standalone.py` dosyasının bir kopyasını oluştur.
2.  Yeni dosyanın adını **`scripts/tune_gemma_model_standalone.py`** olarak değiştir.

#### **Adım 2: Model Oluşturma Mantığını GEMMA'ya Uyarlama**

Yeni oluşturduğun `scripts/tune_gemma_model_standalone.py` dosyası içinde aşağıdaki değişiklikleri yap:

1.  `create_lstm_model` fonksiyonunu bul ve adını `create_gemma_model` olarak değiştir.
2.  Bu yeni `create_gemma_model` fonksiyonunun içindeki tüm LSTM ile ilgili kodları sil.
3.  Fonksiyonun içine, `train_all_models.py`'den referans alarak, bir MLP (Feed-forward) modeli oluşturan yeni bir mantık ekle. Bu fonksiyon, hiperparametreleri (`hidden_size`, `num_layers`, `dropout` vb.) argüman olarak almalıdır.
    *   **Kod Referansı:**
        ```python
        # Bu mantığı yeni create_gemma_model fonksiyonuna uyarla
        
        # ... fonksiyon tanımı ...
        import torch.nn as nn

        class SklearnMLPWrapper:
            def __init__(self, input_size, hidden_size, num_layers, num_classes, dropout, 
                         learning_rate, weight_decay, batch_size, class_weights=None):
                
                # Dinamik MLP modelini oluştur
                layers = []
                layers.append(nn.Linear(input_size, hidden_size))
                layers.append(nn.ReLU())
                layers.append(nn.Dropout(dropout))
                
                for _ in range(num_layers - 1):
                    layers.append(nn.Linear(hidden_size, hidden_size))
                    layers.append(nn.ReLU())
                    layers.append(nn.Dropout(dropout))
                
                layers.append(nn.Linear(hidden_size, num_classes))
                self.model = nn.Sequential(*layers)
                
                # ... (Geri kalan `fit`/`predict` ve early stopping mantığı LSTM versiyonuyla aynı kalabilir) ...
        
        # ... wrapper sınıfının geri kalanı ...
        
        # Fonksiyon, bu wrapper sınıfını return etmelidir
        return SklearnMLPWrapper(...)
        ```

4.  `tune_model` fonksiyonu içindeki `model_factory` tanımını, `create_gemma_model` fonksiyonunu çağıracak şekilde güncelle.

#### **Adım 3: Hiperparametre Arama Uzayını (Search Space) Güncelleme**

`tune_regime_models_standalone.py`'deki `OptunaModelTuner` çağrısı, LSTM'e özel parametreler arar. Bu arama uzayını, GEMMA'nın MLP mimarisine uygun hale getir.

1.  `OptunaModelTuner` sınıfının `_define_search_space_lstm` metodunu kopyalayıp, adını `_define_search_space_gemma` olarak değiştir.
2.  Yeni metodun içinde, LSTM'e özel parametreleri (`weight_decay` gibi) kaldırıp, MLP için anlamlı olan şu parametreleri ve aralıkları tanımla:
    *   `hidden_size`: `trial.suggest_int('hidden_size', 32, 128, log=True)`
    *   `num_layers`: `trial.suggest_int('num_layers', 2, 4)`
    *   `dropout`: `trial.suggest_uniform('dropout', 0.2, 0.6)`
    *   `learning_rate`: `trial.suggest_loguniform('learning_rate', 1e-4, 1e-2)`
    *   `batch_size`: `trial.suggest_categorical('batch_size', [32, 64, 128])`

3.  `OptunaModelTuner`'ın `tune` metodunu, `model_type`'a göre doğru arama uzayını (`_define_search_space_gemma`) seçecek şekilde güncelle.

#### **Adım 4: Argümanları ve Kayıt İşlemlerini Ayarlama**

1.  Script'in `main` bölümündeki `argparse` tanımlarını kontrol et. `--model` argümanının artık `gemma` seçeneğini de kabul ettiğinden emin ol (veya sadece `gemma` için özelleştir).
2.  Sonuçların kaydedildiği `.json` dosyasının adını, `lstm_tuning_` yerine `gemma_tuning_` gibi ayırt edici bir isimle kaydedecek şekilde güncelle (`_save_results` fonksiyonu içinde).

### ✅ Başarı Kriterleri (Success Criteria)

- [ ] `scripts/tune_gemma_model_standalone.py` adında yeni bir dosya oluşturulmuş olmalıdır.
- [ ] Bu script, komut satırından `python scripts/tune_gemma_model_standalone.py` şeklinde çalıştırıldığında hata vermeden başlamalıdır.
- [ ] Script, `data/cache/BTC-USDT_training_data.npz` dosyasını başarıyla okumalıdır.
- [ ] `Optuna` optimizasyon süreci başlamalı ve denemeleri (trials) loglamalıdır.
- [ ] Süreç tamamlandığında, `logs/tuning_results/` altında `gemma_tuning_... .json` adında bir sonuç dosyası oluşturmalıdır. Bu dosya, en iyi `hidden_size`, `num_layers`, `dropout`, `learning_rate` gibi parametreleri içermelidir.

### deliverables Raporlama Şablonu (Reporting Template)

Görevi tamamladığında, aşağıdaki şablonu doldurarak bu issue'ya bir yorum olarak ekle.

````markdown
## Phase 2: GEMMA Tuning Script Creation Report

**Completion Date:** `YYYY-MM-DD HH:MM:SS`
**Executor:** `@github-copilot`

### ✅ Overall Status: `SUCCESS` / `FAILURE`

---

### Task Execution Summary

1.  **New Script Created:**
    - ✅ `scripts/tune_gemma_model_standalone.py` dosyası, `tune_regime_models_standalone.py` şablonu kullanılarak başarıyla oluşturuldu.

2.  **Model Logic Adapted:**
    - ✅ `create_gemma_model` fonksiyonu, `train_all_models.py` referans alınarak bir MLP (Feed-forward) modeli ve `SklearnMLPWrapper` sınıfı içerecek şekilde doğru bir şekilde implemente edildi. LSTM mantığı tamamen kaldırıldı.

3.  **Hyperparameter Space Updated:**
    - ✅ `Optuna` arama uzayı, GEMMA (MLP) için anlamlı olan `hidden_size`, `num_layers`, `dropout`, `learning_rate` gibi parametreleri içerecek şekilde güncellendi.

4.  **Execution and Verification:**
    - ✅ Script, komut satırından `python scripts/tune_gemma_model_standalone.py --trials 10` komutu ile test edildi.
    - ✅ Script, FAZ 1'de oluşturulan `_training_data.npz` dosyasını başarıyla okudu.
    - ✅ `Optuna` optimizasyon süreci başladı ve 10 denemeyi başarıyla tamamladı.
    - ✅ Sonuç olarak, `logs/tuning_results/gemma_tuning_... .json` dosyası, en iyi hiperparametreler ve `balanced_holdout_score` metriği ile birlikte başarıyla oluşturuldu.

---

### Artifacts Produced

- **New Script:**
  - `scripts/tune_gemma_model_standalone.py`
- **Sample Output File:**
  - `logs/tuning_results/gemma_tuning_... .json`

---

### 📝 Sonuç ve Değerlendirme (Conclusion & Analysis)

GEMMA modeli için özel bir hiperparametre optimizasyon aracı başarıyla geliştirilmiştir. Bu araç, projenin mevcut başarılı optimizasyon (`tune_regime_models_standalone.py`) mantığını ve en iyi pratiklerini (veri sızıntısız ölçekleme, `balanced_accuracy` metriği) temel almaktadır.

Bu script'in oluşturulmasıyla, GEMMA modelinin de artık sistematik ve otomatik bir şekilde en iyi performansı verecek şekilde ayarlanabilmesi için gerekli "Ar-Ge" altyapısı tamamlanmıştır.

Sistem, **FAZ 3: Otomasyon Pipeline'ını Kurma** adımına geçmeye hazırdır.
````
